### PR TITLE
Braille Export

### PR DESCRIPTION
**Pull request by andreituicu**
I did not see a specific issue in the tracker, just threads of people asking for it.

This PR adds support for Braille Export native in MuseScore.
Hopefully it will help visually impaired musicians to get easier access to Braille scores.

The majority of the implementation is based on the Music Braille Code 2015 published by the Braille Authority of North America (http://www.brailleauthority.org/music/Music_Braille_Code_2015.pdf). Seemed to be the most comprehensive material on Music Braille Notation that I could find online.

The BRF files should work directly with the Braille display.
For sighted musicians that don't own a Braille display, you can view the Braille characters by opening the exported brf files or the test cases in a text editor (example: Notepad) and set the font to Sim Braille truetype font, it can be downloaded online for free. (This is also how I tested)

The export is marked as experimental in the format drop down, because I would consider this PR to be implementing the MVP of Braille export. There is still room for a lot of improvement and a lot of symbols, conventions and rules that are not yet implemented. I implemented what I personally considered to be the must haves for a big number of scores, but there still a lot that can be done. Braille Music Notation has basically its own layout mechanism and like in the classical music notation there are a lot of edge cases, exceptions, one off rules, score elements being positioned differently when other specific  score elements appear which are most likely not correctly handled by this current implementation.

Biggest chunks of elements that are not supported yet are:
* Percussion instruments
* Lyrics and specific notation for vocal music
* Braille note grouping and stems (current implementation should be readable, however, based on individual note duration)
* Measure split when a measure does not fit into a line
* Others... 🙂 

Disclaimers 🙂 :
* I am not a Braille reader. I simply read chapters from MBC2015 and implemented them.
* I do not own a Braille display, so I never tested that those files work correctly.
* I don't know anything about legal aspects. (example: copying symbols from the book, or using the examples from it as test cases, etc.) If you have any guidance here or preferences, I'm happy to make any changes you recommend, but otherwise I just hope this contribution is useful. 

- [x] I signed [CLA](https://musescore.org/en/cla) <- I signed this back in 2014 (I think), do I need to sign it again?
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/Documentation/blob/master/WorkflowAndGuidelines/CodeGuidelines.md)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [ ] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [ ] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [x] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

File: mtest/CMakeLists.txt
Comment (andreituicu) [inline mtest/CMakeLists.txt:63]: @Jojo-Schmitz : Apologies, I reverted your change here. It was not working for me. The linker was complaining that it cannot find the init functions like (`q_init_resource_mtest`). Unfortunately, I don't have enough knowledge of QT to make a proper fix... 🙁

**Comment by shoogle** [comment]
@andreituicu, thanks so much for this! Could you please change the commit message to "Braille Export collect_artifacts", as that will cause the automated builds to be retained so that we can ask Braillist to download and test. Thanks!

**Comment by MarcSabatella** [comment]
Woah, nice to see you again @andreituicu !  And this is amazing!  Is this something you've been working on independently, or part of some coordinated effort?

Right now I don't have a build environment on "master" working.  In case you aren't aware, "master" is now work towards MuseScore 4, which is likely a ways off.  And maybe that's for the best.  But do you know if it's possible to get this working on the 3.x branch?

**Comment by andreituicu** [comment]
@MarcSabatella This is my quarantine time project. 😸

**Comment by andreituicu** [comment]
@shoogle Sure! Done,

**Comment by MarcSabatella** [comment]
Changing the PR title isn't enough, has to be the commit messages.

Also, the Windows build failed it seems, although I can't seem to discern why.

**Comment by shoogle** [comment]
@andreituicu, you changed the PR title but it needs to be the commit title (`git commit --amend`). It will require a force push.

**Comment by andreituicu** [comment]
@MarcSabatella 
> Right now I don't have a build environment on "master" working. In case you aren't aware, "master" is now work towards MuseScore 4, which is likely a ways off. And maybe that's for the best. But do you know if it's possible to get this working on the 3.x branch?

I've seen something about MuseScore 4 and an announcement about transitioning more to a music composition software. I can certainly try to make it work on MuseScore 3.x branches. If you could please point me to which branch you think would be best for me to try to build on, I can give it a try in the next days and open a PR there as well.

**Comment by andreituicu** [comment]
@shoogle @MarcSabatella done, sorry, I read the comment wrong.

> Also, the Windows build failed it seems, although I can't seem to discern why.

I've developed on Windows, but I've used the MinGW compilation instructions, I could not get QTCreator to work in MSVC and I see that the build is using MSVC, might be related...

**Comment by MarcSabatella** [comment]
3.x would be the branch to try it on.  Realistically, most of the libmscore code hasn't changed, except for the indent/brace style, so chances are good most of your work would apply cleanly.  Officially, 3.x was supposed to be a dead end after 3.5 (aside from any needed bug fixes).  But recently there have been pushes to do some more ambitious engraving improvements and have some of this available before MuseScore 4, so the idea of a MuseScore 3.6 has been floated.  No idea if there have been any real decisions on that or how final they are.  But if that does happen, and if there is any way to get any form of Braille export in at the same time, I will certainly do everything I can to make help make that happen!

This could also be the incentive to see if we can finally get VoiceOver working.  You may have seen, I now have scripts that allow JAWS and Orca to work, but no luck yet with VoiceOver.  Some GSoC work right now might help for MuseScore 4, but I think think it should be possible to tweak what you did way back when and get it working with VoiceOcver.  I've talked to a number of people about this and they all seem to think some combination of picking the right role for the accessible score view object and the right events to fire would make it work, maybe at most needing to also create some dummy child objects to take focus or something.

**Comment by MarcSabatella** [comment]
Oh, and btw - the old IRC channel is defunct, but the current chat is on https://t.me/musescoreeditorchat

**Comment by shoogle** [comment]
@andreituicu, also feel free to join the MuseScore Developers chat on Telegram https://t.me/musescoreeditorchat

File: mu4/domain/importexport/internal/braille/exportbraille.cpp
Comment (shoogle) [inline mu4/domain/importexport/internal/braille/exportbraille.cpp:585]: These lines cause build errors with MSVC. See https://stackoverflow.com/a/33423538

Comment (shoogle) [reply mu4/domain/importexport/internal/braille/exportbraille.cpp:585]: Probably best to use a `std::vector` of QStrings.

Comment (andreituicu) [reply mu4/domain/importexport/internal/braille/exportbraille.cpp:585]: @shoogle thank you for the pointer! I'll convert them into dynamic calls. MinGW was happy with them.

Comment (Jojo-Schmitz) [reply mu4/domain/importexport/internal/braille/exportbraille.cpp:585]: It indeed does not

File: mtest/CMakeLists.txt
Comment (Jojo-Schmitz) [reply mtest/CMakeLists.txt:63]: Which Qt version are you using? MuseScore officislly still uses 5.9, but I'm building with 5.12 and have no issues with this

File: mtest/README.md
Comment (Jojo-Schmitz) [inline mtest/README.md:9]: Wouln't that break the MINGW builds?

File: Makefile.mingw
Comment (Jojo-Schmitz) [inline Makefile.mingw:23]: unnecessary change.

File: libmscore/note.h
Comment (Jojo-Schmitz) [inline libmscore/note.h:556]: unnecessary change

File: mscore/file.cpp
Comment (Jojo-Schmitz) [inline mscore/file.cpp:1783]: unnecessary change

File: mtest/README.md
Comment (Jojo-Schmitz) [inline mtest/README.md:28]: `Windows MinGW` is a tautology, the W in MinWG stands for Windows ;-)

File: mtest/braille/CMakeLists.txt
Comment (Jojo-Schmitz) [inline mtest/braille/CMakeLists.txt:5]: Not for new files anymore, but `2020 MusEscore BVBA`

File: mu4/domain/importexport/internal/braille/exportbraille.cpp
Comment (Jojo-Schmitz) [reply mu4/domain/importexport/internal/braille/exportbraille.cpp:585]: Search the source for `VLA` to find how this is being worked around. Example:

    #if (!defined (_MSCVER) && !defined (_MSC_VER))
          char txt[size+1];
    #else
          // MSVC does not support VLA. Replace with std::vector. If profiling determines that the
          //    heap allocation is slow, an optimization might be used.
          std::vector<char> vtxt(size+1);
          char* txt = vtxt.data();
    #endif

File: mtest/CMakeLists.txt
Comment (andreituicu) [reply mtest/CMakeLists.txt:63]: I am building with Qt 5.12.9  for MinGW on 64 bits.

File: mtest/README.md
Comment (andreituicu) [reply mtest/README.md:9]: The top level `Makefile.mingw` still needs to be invoked and that one works. However, in the resulting build directory, under the mtest folder there is no `Makefile.mingw` generated, the file is simply called `Makefile`.

File: mtest/CMakeLists.txt
Comment (Jojo-Schmitz) [reply mtest/CMakeLists.txt:63]: Me too, and it works

Comment (andreituicu) [reply mtest/CMakeLists.txt:63]: The only thing I can think of that might be related is: Have you tried to remove all build files and start a build from scratch? If I understand correctly, the `.qrc` files are missing from the build directory and if you have them from a previous build then it would work. Unfortunately, like I said I don't know much here so it may be that if you try to delete all build files you will just wait an hour to rebuild everything and it will work for you and it is just me doing something wrong. What's more weird to me is that for the build Cmake file I didn't have to revert it, but just for the mtest CMake...

File: Makefile.mingw
Comment (andreituicu) [reply Makefile.mingw]: Addressed in b0365cb. Thank you! 👍

File: libmscore/note.h
Comment (andreituicu) [reply libmscore/note.h]: Addressed in b0365cb. Thank you! 👍

File: mscore/file.cpp
Comment (andreituicu) [reply mscore/file.cpp]: Addressed in b0365cb. Thank you! 👍

File: mtest/README.md
Comment (andreituicu) [reply mtest/README.md]: Addressed in b0365cb. Thank you! 👍

File: mtest/braille/CMakeLists.txt
Comment (andreituicu) [reply mtest/braille/CMakeLists.txt]: Addressed in b0365cb. Thank you! 👍

File: mu4/domain/importexport/internal/braille/exportbraille.cpp
Comment (andreituicu) [reply mu4/domain/importexport/internal/braille/exportbraille.cpp]: Addressed in b0365cb. Thank you! 👍 I used directly the `std::vector<QString> line(nrStaves);` to not have diverging code depending on the compiler if one code can accommodate all. Let's see what the build says.

**Comment by shoogle** [comment]
I have contacted three experts and sent them Braille output for these two scores:

- Beethoven's *[Für Elise](https://musescore.com/openscore/scores/4053926)*.
- Mahler's *Kindertotenlieder* ([1st movement](https://musescore.com/openscore-lieder-corpus/scores/5054519)).

It's the weekend, so I've only heard back from one person so far. That person said that the octave signs have been used incorrectly, though they didn't say how.

I also noticed the following in my own review:

- Lyrics are not included (perhaps intentionally at this stage?)
- Some text from the cover pages of Für Elise is included, but most text is not.
    - It looks like you try to get the key information (like Title, Composer, Copyright, etc.) but there is some other text included.
    - Ideally there would be an option to include all text from text frames (e.g. in Preferences> Advanced Preferences) and this should be enabled by default since a blind person can easily edit the file to delete it if they don't want it. (I'm thinking Braille output on MuseScore.com here, which should include everything by default.)

**Comment by andreituicu** [comment]
@shoogle 
> I have contacted three experts and sent them Braille output for these two scores:

That is fantastic,, thank you! I don't know anybody who knows Braille. Up until now it was just me reading the spec and writing the code based on it. I am very interested to hear what people who actually know Braille think about it.

> That person said that the octave signs have been used incorrectly, though they didn't say how.

I am very interested to hear what is incorrect. I know that I didn't implement octave marks 100%
for chords you can see the TODO here: https://github.com/musescore/MuseScore/pull/6343/files#diff-dc87209ebc371bc80e99d511934f84eeR1297 and for individual notes, I simplified a bit this initial implementation. I put the octave mark at the beginning of the line, after every symbol where the book said needs an octave mark and **every time a note is in different octave then the previous specified octave**. The last part means that some octave marks might be redundant, because according to the spec, the octave mark is added normally only when **the previous note is in a different octave and the interval is higher then a 5th.** I thought this would be a harmless simplification for a first implementation because it means there are some extra octave marks, no missing ones. But let's see what they say, maybe there's more.

> Lyrics are not included (perhaps intentionally at this stage?)

Yes, lyrics are not included. This is also specified in the PR description.

> It looks like you try to get the key information (like Title, Composer, Copyright, etc.) but there is some other text included, including the composer name twice.

Yes, here I took the implementation from musicxml export. https://github.com/musescore/MuseScore/blob/master/mu4/domain/importexport/internal/musicxml/exportxml.cpp#L5459

> Ideally there would be an option to include all text from text frames (e.g. in Preferences> Advanced Preferences) and this should be enabled by default since a blind person can easily edit the file to delete it if they don't want it. (I'm thinking Braille output on MuseScore.com here, which should include everything by default.)

I agree here, but I intentionally didn't go yet into preferences and settings into this first PR. There are more settings that would be of interest for Braille exports:
* What text to be included/excluded
* How many characters on a line (currently hard coded to `40`)
* Facsimle/Nonfacsimle (Nonfacsimle : export for a Braille reader that is only interested in a pure Braille rendition, without extras. Facsimle: Include elements in the rendition that normally don't appear in a Braille score, but appear in a print score. Current implementation does not support nonfacsimle renditions)
* Probably a lot more other settings
* Contractions in text: on/off (contractions not supported yet)

Current implementation exports what a simple score usually has: the title (first vertical frame), the copyright owner, the composer, lyricist, arranger, poet, translator.

Just to set the expectations when looking at this PR. It is by no means the definitive Braille export, finished and done. There is a ton more to do (lots of elements to be exported, lots of simplifications on my part, lots things I probably don't even know I missed). There are a lot of chapters of the spec, I didn't even get to read yet, or skipped intentionally because they were optional or not frequently used in scores. This PR implements the basics, what I personally believed would be required to have something useful and the point where I believed that it is time to ask for feedback and start integrating back upstream to continue with incremental PRs (and hopefully get other people involved too, with stories in the tracker 😊). This is why the export is also marked as `experimental`. 

This doesn't mean I'm not looking for feedback, I am looking forward to any kind of feedback from you, other contributors, users, Braille experts, etc, so please keep it coming and thank you very much! 🙂 Maybe we have to structure it in:
* what is blocking and needed for this first implementation. I am happy to address anything here.
* what is an improvement that can go into the tracker as a feature/story/epic that can be added in dedicated PRs.

Please let me know what you think and/or if you don't agree! This is just my current thinking. Happy to do it any other way!

**Comment by shoogle** [comment]
Some more feedback for you:

> Wow! This export plugin is basically good, with perfect bar-over-bar
formatting. Although it has lots of limitations, this can be a very
good starting point. Ok, now I give my test result:
> 1. The tool is a good starting point, although very basic, without
this and that;
> 2. It has bar-over-bar  format, aligned every bar well;
> 3. It seems only system texts are translated;
> 4. No foreign language translation, so accent letters and unicode
letters like Russian is exported as they are;
> 5. No part names and abbreviations, no piano hand signs;
> 6. Slur definitions are very basic;
> 7. lots of rules missing, but can be quickly implemented;
> 8. Empty staves can't be hidden;

I imagine this is stuff that you are already aware of since it refers to limitations (i.e. things you are not even trying to do). I have asked the testers to provide details about actual problems with what's there rather than pointing out things that are missing.

**Comment by shoogle** [comment]
Regarding octave signs:

> Below is an explanation of the use of octave signs (markings) but it is more involved as you have to take into account the format being used so it is much more involved than just reading the basic rules.
>
> In essence, an octave is said to be calculated from for instance the lowest C on the piano to the B above, that being called the First Octave. Middle C is the start of the Fourth Octave on the piano.
>
> Basic rules are:
> - The first note of a piece of music, or main part or section requires an octave sign.
> - If the next note is an ascending or descending second or third, no octave sign is needed even if those notes are in a different octave. If a dynamic mark comes between these notes then an octave marking will be required.
> - If the next note forms an interval of an ascending or descending fourth or fifth, it only needs an octave sign if the note is in a different octave.
> - Intervals of a sixth or greater will need octave signs, regardless.
>
> As already stated, there are plenty of exceptions so the text above is the basic position.

I'm not sure how helpful that is to you, but it was accompanied by the offer of a meeting via Zoom or Microsoft Teams to explain things in more detail. I will send info via Telegram.

**Comment by shoogle** [comment]
More feedback:

> I just tested the braille module again using Mozart's 12 variations.
This easy piece aborts the export at variation 6, which I don't know
the reason. Mscz file attached.
>
> Report:
> 1. All frames or texts after the title are not exported.
>  2. Braille systems are not broken according to print systems or
double bars. This is optional, but should be used when a frame is met,
e.g., variation 2 or the like, if the text is entered as tempo or
system text or frame.
> 3. For octave marks, the real problem is, no matter if the adjacent
notes reach the limit, notes at different octaves are added with
octave marks. In braille, octave mark should be added if the second
note in another octave is a fourth far from the current note,
otherwise no octave mark is needed.
> 4. Left hand in-accord sign should first introduce lower voice, then
upper voice.
> 5. If Musescore has measure reset for each variation, braille should
also reset measure to 1 or 0 (if pickup).
>
> Correct braille attached. Note that it's a edited version, so there
may be differences from the Musescore source. But the general
appearance can be learned.

Attachment: [Zwoelf_Variationen_in_C.zip](https://github.com/musescore/MuseScore/files/4947486/Zwoelf_Variationen_in_C.zip)

**Comment by shoogle** [comment]
Even more feedback:

> I've just taken a quick look at the C major scale. I hope the following comments are helpful:
>
> Original output (just the part that counts):
> ```
> a >/l#d4 "?:$] \[W.?<K
> ```
>
> Ideal output should be:
>```
>    #d4
>a >/l"?:$] \[W?<K
>```
>
> Comments:
>
> 1. CR+LF problem. Windows files should have CR+LF at the end of each line. Some programs may not read the file correctly with only LF.

My advice is to always use CR+LF, even on Linux and Mac. They don't complain like Windows does, and it makes it easier to provide Braille on MuseScore.com. Braille files on http://www.brailleorch.org/ use CR+LF.

> 2. Key and time signature are, in braille, normally made to stand on a line by themselves, with any initial expression markings for the piece. The line is often centred. Note this means a change of order of the signs between print and braille. If a key change or time sig change occurs mid-piece, this can be written in-line, with spaces around; the clef sign always is the last thing. (I'm sure much more can be said on this one, and I would suggest it is not critical at this stage).
> 3. All the basic signs in the braille music are basically correct.
> 4. Octave sign problem. There is no need to show the fifth octave sign before the last note, because the interval is only a second (semitone). I would suggest the rules for showing (or not showing) octave signs are critical. basically you look at the interval between the current note and the next.
> If it is up to and including a third, you never need an octave sign;
> if it is a sixth or greater, you always need an octave sign;
> For a fourth or fifth, you need an octave sign if the next note is in a different octave.
> This third rule is the hardest to understand.
>  
> Example:
>
> Fourth octave C followed by fourth octave G.
> Interval: fifth.
> Octave sign needed before the second note: no, because they are in the same octave.
> Braille: `"?\`
>
> However:
> Fourth octave G followed by Fifth octave D
> Interval: still a fifth
> Need octave sign: yes, because the second note is in a different octave
> Braille: `"\.:`
>
> If this second example did not have the octave sign, a reader would interpret it as fourth octave G followed by fourth octave C (interval, a fourth).

Attachment: [C_Major.zip](https://github.com/musescore/MuseScore/files/4948179/C_Major.zip) (Note that the Braille is automated output for the MSCZ using your code.)

**Comment by shoogle** [comment]
Yet more feedback:

> May I report another fairly significant problem for the [Fur Elise](https://musescore.com/openscore/scores/4053926): It is convention for piano score that the right hand are marked at the beginning of each line with the sign (dots 46, 345) (braille characters `.>` ) or `⠨⠜`. Similarly, the left hand uses dots 456, 345.
>
> Thus for example, picking a few lines in the score: Original reads: (bar 38)
> ```
> ch .f'").&%z f'")c.&%zc &%zc&zc&zc
>   m         m          m
> da .&%z&")*.zy "imy&!  jm&%()
>   m           ^!_&!mx ^&_&%(mx
> ```
>
> Should read:
>
> ```
> ch .>.f'").&%z f'")c.&%zc &%zc&zc&zc
>   _>'m        m          m
>da .>.&%z&")*.zy "imy&!  jm&%()
>   _>'m          ^!_&!mx ^&_&%(mx
> ```
>
> A little point of interest, I inserted a dot 3 (') in the left hands to separate the dots 345 from the next sign (M) as it contains left-hand dots. (I don't want to bore you too much with the intricacies of the braille music code rules).
>
> Another point of interest in bar 41 (above), left hand, there is possibly the need for "tracker dots" a line of '''' to cover the large gap between the M and the next bar.
>
> Another very important point, possibly for a little later, is interval direction. I was very pleased to see this is generally correct, but the important point is that the interval direction in piano music is normally by the hand, not the clef.
So, if the left hand moves to treble clef, then there is a variant sign (345, 34, 13) to show interval direction is still up
Interval direction is a complex debate and I'll not go into further detail now..
>
> I have to say, for a first time look at braille music produced from a new system, this is pretty good, though some improvements will make it really nice.

**Comment by kuwitt** [comment]
Forgive my intrusion as an average user, who don't have any coding skills: But I am very impressed with this PR and the discussion about it and I very much hope that it will be successfully merged in a future release.

File: mu4/domain/importexport/internal/braille/exportbraille.cpp
Comment (Marr11317) [inline mu4/domain/importexport/internal/braille/exportbraille.cpp:318]: coversion: typo?

File: mu4/domain/importexport/internal/braille/exportbraille.cpp
Comment (Marr11317) [inline mu4/domain/importexport/internal/braille/exportbraille.cpp:438]: It indeed seems more logical to me to have them in libmscore.

File: mu4/domain/importexport/internal/braille/exportbraille.cpp
Comment (Marr11317) [inline mu4/domain/importexport/internal/braille/exportbraille.cpp:554]: why not save all metatags?

**Comment by andreituicu** [comment]
The Braille octave mark implementation should be correct in the latest commit according to the Braille rules for melody line and chords. (There are still some small differences from some of the examples in MBC2015, but I didn't find any written rule or consistent pattern to explain them).

**Comment by Jojo-Schmitz** [comment]
rebase needed

**Comment by MarcSabatella** [comment]
FWIW< see comments from a user here: https://musescore.org/en/node/307922#comment-1034592 - they are impressed overall, just a few small tweaks they'd like.

**Comment by Jojo-Schmitz** [comment]
And another rebase is needed

**Comment by andreituicu** [comment]
@Jojo-Schmitz rebased on current `master`. Re-run the tests locally, everything looks good and the build checks look green. Thank you!

@MarcSabatella Thank you very much! I'm happy that people find this useful! 🙂

> 1) The measure number printed on every line should be preceded by the number sign (dot3456)

Yes, I'm aware that some scores write the measure numbers like this. I've seen both variants. Just for clarity:
* current implementation writes the measure numbers as: `a` - measure `1`, `aj` - measure `10`
* the proposed feedback and probably more correct according to the Braille rules is numbers are prefixed by `#`, so: `#a` - measure `1`, `#aj` - measure `10`.

The reason I personally liked the first variant and I implemented in MuseScore is because, there are only 40 characters in the Braille line and the measure numbers are taking at minimum already at least 2 of them (number + space), so I didn't liked the idea of taking one more character on each line for `#` if it is already clear that those are measure numbers.

> 2) If the first measure is a complete measure, the measure number is one. If the melody starts with an incomplete measure the measure number should be zero. The measure numbers then correspond to the normal printed score.

I wasn't aware of this, unfortunately. Is using the `Measure#isAnacrusis()` method the correct way to go about implementing this? This is your decision, I wouldn't wait on merging the PR for this. If I find how to implement this properly before it is merged, I'll push the change here directly. Otherwise, I'll open a new PR.

> 3) Words such as "dolce" and "sotto voce" are not printed in the braille translation. The text should be preceded by word sign (dot345). The first note following any use of word sign should be preceded by an octave sign, which will indicate the reestablishment of the music code.

I think this might be expected with the current implementation. I'll have to look into the score they used, but most text elements are not exported. I was waiting for this PR to be merged to start working on extending the implementation to more text elements.

**Comment by MarcSabatella** [comment]
@andreituicu thanks for rebasing!

Regarding the leading "#" for measure numbers, I would suggest keeping to the standard.  Maybe leave a TODO comment about someday adding a setting to control this.  I could imagine there eventually being a whole group of settings to control aspects of the Braille export, maybe a separate page in the Style dialog, or in Preferences.

Regarding the measure numbers themselves, you shouldn't need to be checking the anacrusis status yourself, and in fact that wouldn't be sufficient, as there are other cases where the printed measure can differ from the internal.  Instead you should simply use the same information that is used for the printed number.  Looks like you are currently tracking measure numbers yourself by incrementing an index and managing the mmrests manually, but instead, you simply query the measure for its printed number.  I'm not exactly sure how this is managed as I think there were changes pretty recently, but probably Measure::noText() is what you want.

**Comment by andreituicu** [comment]
@MarcSabatella: `MeasureNumber` was the first place I looked. Unfortunately, I didn't find a reliable way to make it work, so I settled on using a dedicated index as a temporary fix to finish this first implementation.

**Comment by MarcSabatella** [comment]
Looking further at the code, I think actually it's just MeasureBase::no() + 1.  I had been thinking this was the internal number only, but it seems it isn't, _no is constantly adjusted during layout according to measure properties.

**Comment by andreituicu** [comment]
Thank you! I'm looking at the same function now, I'm building and trying it out. I find it a bit weird, because I don't understand how it could get to be `-1` so an incomplete measure at the beginning would get `no() + 1 == 0`, but let's see how it works.

**Comment by MarcSabatella** [comment]
Well, it would be numbered 1, but so would the first *full* measure, assuming the incomplete measure had the "Exclude from measure count" property set, which it normally should.  Similarly, if the user sets the "Add to measure number" property on any given measure, this also gets reflected in no().  So it really should guarantee you are exporting the same number that is actually displayed in the score, which is the whole point.

So you might want to special case a measure at the beginning of the score that has the "exclude from measure count" property set, and explicitly export that as 0.  Other than that, though, no() + 1 should give the correct result.

**Comment by andreituicu** [comment]
@MarcSabatella Thanks a lot! I did a few tests and in all of them `no() + 1` worked so I pushed the change in 
a560040! All my existing mtests were still passing.🙂

**Comment by Jojo-Schmitz** [comment]
That `collect_artifacts` is no longer needed, these are generated now in any case.
Having 3 commits in this PR with an identical commit title is not good, please either give them meaningful individual titles or squash them into one commit.

**Review by Jojo-Schmitz — APPROVED** [review]

**Comment by andreituicu** [comment]
@Jojo-Schmitz ✅ Done. Thank you!

**Pull request merged**
